src/BaseballGame.js

```diff
@@ -0,0 +1,74 @@
+import { MissionUtils } from "@woowacourse/mission-utils";
+import { generateRandomNum } from "./utils/generateRandomNum";
+import { isInputValid, isEndInputValid } from "./utils/validation";
+import { getResult, printResult } from "./utils/getResult";
+
+class BaseballGame {
+  #COMPUTER_NUM_LEN = 3;
+  #ComputerNums;
+  #UserNums;
+
+  setComputerNums(nums) {
+    this.#ComputerNums = nums;
+  }
+
+  setUserNums(nums) {
+    isInputValid(nums, this.#COMPUTER_NUM_LEN);
+    this.#UserNums = nums.split("").map((str) => Number(str));
+  }
+
+  /**
+   * 게임의 전체 진행을 위한 메서드입니다.
+   */
+  async runGame() {
+    await this.startGame();
+    await this.playGame();
+    await this.endGame();
+  }
+
+  /**
+   * 게임 시작을 위한 메서드입니다.
+   * 게임 시작 멘트를 출력하며, 컴퓨터의 숫자 3개를 정합니다.
+   */
+  async startGame() {
+    await MissionUtils.Console.print("숫자 야구 게임을 시작합니다.");
+    this.setComputerNums(await generateRandomNum(1, 9, this.#COMPUTER_NUM_LEN));
+  }
+
+  /**
+   * 게임 플레이를 위한 메서드입니다.
+   * 값을 입력 받고, 받은 값의 유효성 체크와 결과를 출력합니다.
+   */
+  async playGame() {
+    let isUserWin = false;
+    while (!isUserWin) {
+      const USER_INPUT = await MissionUtils.Console.readLineAsync("숫자를 입력해주세요 : ");
+
+      this.setUserNums(USER_INPUT);
+
+      const { strike, ball } = await getResult({
+        computerNums: this.#ComputerNums,
+        userNums: this.#UserNums,
+      });
+      await printResult({ strike, ball });
+      if (strike === 3) {
+        await MissionUtils.Console.print("3개의 숫자를 모두 맞히셨습니다! 게임 종료");
+        isUserWin = true;
+      }
+    }
+  }
+
+  /**
+   * 게임 종료를 위한 메서드입니다.
+   * 재시작과 완전한 종료로 나뉩니다.
+   */
+  async endGame() {
+    const USER_INPUT = await MissionUtils.Console.readLineAsync(
+      "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요. \n"
+    );
+    isEndInputValid(USER_INPUT);
+    if (USER_INPUT === "1") await this.runGame();
+    else await MissionUtils.Console.print("게임이 종료되었습니다.");
```

Comment:
```suggestion
    if (USER_INPUT === "1"){ 
       await this.runGame();
       return;
    }
    
    MissionUtils.Console.print("게임이 종료되었습니다.");
```
저는 이런 경우에 `early return`을 사용하려고 하는 것 같아요! 이건 취향 차이일 것 같네요🙃
그리고 `MissionUtils.Console.print` 메서드도 `await`에 영향이 가지 않을 것 같아요

Reply:
오 여기에 `return`을 넣으시는 이유가 뭔가용? :thinking:

Reply:
`else`를 생략할 수 있어서 선호하는 것 같습니다. early return을 사용하면 `depth`가 늘어나지 않아서 좋은 것 같아요!

Reply:
혹시 `return`없이도 `else`는 생략 가능하지 않나요? 지금 이 경우에서는 depth가 늘어나지도 않구요! 
``` suggestion
    if (USER_INPUT === "1") await this.runGame();
    await MissionUtils.Console.print("게임이 종료되었습니다.");
```
오히려 이렇게 표현해도 되지 않을까요?
`return`을 적어줘야 함수 탈출이 명확하려나요? :thinking:

Reply:
> 혹시 `return`없이도 `else`는 생략 가능하지 않나요? 지금 이 경우에서는 depth가 늘어나지도 않구요!

-> 맞습니다! 지금 코드에서 early return의 이점은 `else`문 생략일 것 같은데, 추후에 요구사항이 늘어나서 입력 값이 `2`이고 `3번 안에 3스트라이크를 맞춘 경우`와 같이 조건이 늘어나면 depth가 늘어날 요소가 생길 것 같아서 early return을 사용하면 유지보수하기 편할 것 같아요!

> 오히려 이렇게 표현해도 되지 않을까요? `return`을 적어줘야 함수 탈출이 명확하려나요? 🤔
return을 적어주지 않으면 첫 번째 게임이 끝나지 않은 상태로 두 번째 게임이 시작되어서 게임 종료될 때 종료 문구가 중첩되어서 나올 것 같아요! 그래서 return으로 첫 번째 게임을 끝내주고 두 번째 게임으로 들어가야 할 것 같아요

이렇게 적으면 조금 이해하기 어려우실 것 같아서 추가 설명 드릴게요! 혹시라도 이해가 잘 안가시면 더 물어보셔도 됩니다!!
```
초기 게임 시작(첫 번째 게임 시작)
3스트라이크 
유저 입력: 1 
게임 재시작(두 번째 게임 시작, 아직 첫 번째 게임은 끝나지 않은 상태 await 걸려있기 때문에  뒤에 게임이 끝나야 끝남)
3스트라이크
유저 입력: 2 (두 번째 게임 종료)
게임이 종료되었습니다. (두 번째 게임 종료 문구 출력)
게임이 종료되었습니다. (첫 번째 게임 종료 문구 출력)
```

Reply:
추가적으로 지금 코드를 보다가 알았는데 예지님 코드에서 유효성 검사하는 부분의 인터페이스를 봤을 때 유효성 검사에 통과하지 못하면 어떤 일이 발생하는지 파악하려면 내부 코드에 들어가야 알 수 있는 것 같아요!
```js
  async endGame() {
    const USER_INPUT = await MissionUtils.Console.readLineAsync(
      "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요. \n"
    );
    
   // 1. 함수 이름은 유효성 검사를 하는 것 같은데 통과하지 못하면 어떻게 되는지 여기서는 파악하기 어렵다
    isEndInputValid(USER_INPUT);  // 기존 코드
    
   
   // 2. is, has, should 는 주로 boolean 값을 다룰 때 사용하는 네이밍인 것 같아서 내부에서 에러를 throw하지 않고 boolean 반환
   // 3. 반환받은 boolean 값으로 에러를 throw 하면 endGame함수 인터페이스만 보고도 동작을 알 수 있다!
   const isEndInputValidate = isEndInputValid(USER_INPUT); 
   
   
   if(isEndInputValidate){
     throw new Error("에러 발생 ㅇ_ㅇ!")
   }

   // another code...
  }
```
이렇게 유효성 검사 함수에서 `boolean` 값을 반환받고 그 값에 따라 에러를 `throw`하는 로직은 어떻게 생각하시나요?!

Reply:
아 그렇네요..! 당연한건데 생각지도 못했어요..! `return`을 써주는게 맞겠네요! 감사합니다 :+1: 
유효성 검사 관련된 내용도 리팩토링 하면서 더 고민해보겠습니다! 감사해요 :relaxed: